### ISSUE TITLE
tf/keras model.fit training accuracy doesn't match model.predict() accuracy on training data despite having only one training batch

### ISSUE THREAD

**Issue by justinmulli** [tensorflow/tensorflow#38486]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Mac OS Catalina
- TensorFlow installed from (source or
binary): I tried both - The issue persists regardless
- TensorFlow version (use command below):'2.1.0'
keras version '2.2.4-tf'
- Python version: 3.7.7

**Describe the current behavior**
When I fine tune a pre-trained resnet using tensorflow.keras.applications.ResNet50, the accuracy shown in training does not match the accuracy of using predict on the same training data, despite the batch size being the size of the entire training set (i.e. only one batch per epoch).

I did my due diligence researching this problem and apparently it has something to do with the batch normalization - but surely this is an error that needs to be fixed?

**Describe the expected behavior**
I expect the training accuracy during training after calling model.fit to be the same as the accuracy using model.predict on the same data, given that the batch size is the size of the entire training set.

**Standalone code to reproduce the issue** 
Here's the data https://www.icloud.com/iclouddrive/0xyiKQ2D7CuucbbQ-M8NSPilQ#tf
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
from tensorflow.keras.layers import Conv2D, Input, MaxPool2D, add, Flatten, Dense
import tensorflow.keras as keras
import tensorflow as tf
import numpy as np
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import confusion_matrix
from seaborn import heatmap
tf.keras.backend.clear_session()from tensorflow.keras.applications import ResNet50
from tensorflow.keras.models import Sequential
import os


os.chdir('/users/justusmulli/bayer')
resnet_weights_path = 'resnet50/resnet50_weights_tf_dim_ordering_tf_kernels_notop.h5'
os.chdir('/users/jm')


def classification_sampler(x, y, distribution = 'balanced'):
    counts = np.unique(y, return_counts = True)
    class_size = {}
    s = np.argmin(counts[1])
    if distribution == 'balanced':
        for i in range(len(counts[0])):
            class_size[counts[0][i]] = counts[1][s]
    else:
        for class_name in distribution:
            if distribution[class_name]>1:
                class_size[class_name] = distribution[class_name]
            else:
                class_size[class_name] = counts[1][s]*(distribution[class_name]/class_size[counts[0][s]])
    x_sample = []
    y_sample = []
    for i in range(len(class_size)):
        class_name = counts[0][i]
        x_sample.extend(x[y==class_name][:class_size[class_name]])
        y_sample.extend(y[y==class_name][:class_size[class_name]])
    return np.array(x_sample), np.array(y_sample)


x_train = np.load('x_train.npy')
y_train = np.load('y_train.npy')
x_val = np.load('x_test.npy')
y_val = np.load('y_test.npy')
x_train, y_train = classification_sampler(x_train, y_train)
x_val, y_val = classification_sampler(x_val, y_val)
encoder = OneHotEncoder()
encoder.fit(y_train.reshape(-1,1))
y_train = encoder.transform(y_train.reshape(-1,1)).toarray()
y_val = encoder.transform(y_val.reshape(-1,1)).toarray()

model = Sequential()


model.add(ResNet50(include_top = False, pooling = 'avg'))

model.add(Dense(512, activation = 'relu'))
model.add(Dense(128, activation = 'relu'))

model.add(Dense(3, activation = 'softmax'))

model.layers[0].trainable = False

from tensorflow.python.keras import optimizers
optimizer = keras.optimizers.Adam(learning_rate = 0.00000005)
model.compile(optimizer = optimizer, loss = 'categorical_crossentropy', metrics = ['accuracy'])
es = keras.callbacks.EarlyStopping(monitor='val_accuracy',
                                  min_delta=0,
                                  patience=100,
                                  verbose=0, mode='max')
hist = model.fit(x_train, y_train, validation_data = (x_val, y_val), batch_size = 267, epochs = 1000, callbacks = [es])
b = hist.model.predict(x_train)
accuracy_score(np.argmax(b, axis = 1), np.argmax(y_train, axis = 1))
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
Epoch 101/1000
267/267 [==============================] - 14s 52ms/sample - loss: 1.1680 - accuracy: 0.3708 - val_loss: 1.1356 - val_accuracy: 0.3333

b = hist.model.predict(x_train)
accuracy_score(np.argmax(b, axis = 1), np.argmax(y_train, axis = 1))
0.3333333333333333

**Comment by khimraj** [comment]
@justinmulli, when you use `fit`, at each batch of the training data the weights are updated. The accuracy value returned by the `fit` method is not the mean of the accuracy of the final model, but the mean of the accuracy of all slightly different models used on each batch.
On the other hand, when you use `predict`, the same model is used on the whole dataset. And this model actually doesn't even appear in the accuracy of the fit method since even at the last batch of training, the loss computed is used to update the model's weights.

**Comment by justinmulli** [comment]
> @justinmulli, One of reason is that regularization mechanisms, such as Dropout and L1/L2 weight regularization, are turned off at testing time and also when you use `fit`, at each batch of the training data the weights are updated. The accuracy value returned by the `fit` method is not the mean of the accuracy of the final model, but the mean of the accuracy of all slightly different models used on each batch.
> On the other hand, when you use `predict`, the same model is used on the whole dataset. And this model actually doesn't even appear in the accuracy of the fit method since even at the last batch of training, the loss computed is used to update the model's weights.
> Reference -https://keras.io/getting-started/faq/#why-is-the-training-loss-much-higher-than-the-testing-loss

Ok so I shouldn't be getting the accuracy displayed in the last epoch, but I get 0.33333 accuracy using predict() no matter how many epochs I train it for, so something is wrong. the past 50 epochs did not have an accuracy of 0.3333333 - model.predict always predicts the same class and gets 0.333333 accuracy no matter when it is stopped or what the training accuracy is.

**Comment by gowthamkpr** [comment]
@justinmulli There is not much difference between training, validation and prediction accuracy. Lets say if your training and validation accuracy are significantly different from prediction accuracy then there is a problem. 
Looks like there are many factors that can contribute to you facing this issue. Please go through this [issue](https://stackoverflow.com/questions/41488279/neural-network-always-predicts-the-same-class). Also I would recommend you to post this question on stackoverflow where there is a wider community to respond. Thanks!

**Comment by justinmulli** [comment]
> @justinmulli There is not much difference between training, validation and prediction accuracy. Lets say if your training and validation accuracy are significantly different from prediction accuracy then there is a problem.
> Looks like there are many factors that can contribute to you facing this issue. Please go through this [issue](https://stackoverflow.com/questions/41488279/neural-network-always-predicts-the-same-class). Also I would recommend you to post this question on stackoverflow where there is a wider community to respond. Thanks!

Shouldn't the training accuracy printed during training (at least somewhat) reflect the accuracy on the training data using model.predict()?

**Comment by gowthamkpr** [comment]
@justinmulli This is not a bug, its intended behaviour. I would recommend you to post this in stackoverflow as there is a wider community to answer. Thanks!

**Comment by gowthamkpr** [comment]
Closing this issue as it has been inactive for a long time. Please add additional comments or open a new issue. Thanks!
